src/pynxtools_xps/kratos/metadata_kratos.py

```diff
@@ -0,0 +1,350 @@
+        """
+        Extract units for the metadata containing unit information.
+
+        Example:
+            analyser_work_function: 4.506eV
+            -> analyser_work_function: 4.506,
+               analyser_work_function_units: eV,
+
+        Parameters
+        ----------
+        key : str
+            Key of the associated value.
+        value : str
+            Combined unit and value information.
+
+        Returns
+        -------
+        value :
+            value with units.
+        unit : str
+            Associated unit.
+
+        """
+
+        pattern = re.compile(r"([-+]?\d*\.?\d+(?:[eE][-+]?\d+)?)([a-zA-Z]+)")
+        match = pattern.match(value)
+
+        if match:
+            value, unit = match.groups()
+        else:
+            unit = ""
+
+        unit = convert_units(unit)
+
+        if key in UNIT_MISSING:
+            unit = UNIT_MISSING[key]
+
+        return value, unit
+
+    def map_values(self, key: str, value, field_type):
+        """
+        Map values to corresponding structure and field type.
+
+        Parameters
+        ----------
+        key : str
+            KratosMetadata dataclass key.
+        value : str
+            Original value.
+        field_type : class
+            Class to be mapped onto.
+
+        Returns
+        -------
+        value
+            Value of correct type and internal structure.
+
+        """
+
+        value_function_map: Dict[str, Any] = {
+            "date_created": _parse_datetime,
+            "description": _convert_description,
+            "charge_neutraliser": convert_bool,
+            "deflection": _convert_xray_deflection,
+        }
+
+        if key in value_function_map:
+            map_fn = value_function_map[key]
+            value = map_fn(value)
+        return field_type(value)
+
+    def flatten_metadata(self):
+        """
+        Flatten metadata dict so that key-value pairs of nested
+        dictionaries are at the top level.
+
+        Parameters
+        ----------
+        metadata_dict : dict
+            Metadata dict with KratosMetadata fields as keys.
+
+        Returns
+        -------
+        flattened_dict : dict
+            Flatted metadata_dict without any nested substructure.
+
+        """
+
+        flattened_dict = {}
+
+        for key, value in self.metadata.dict().items():
+            if isinstance(value, dict):
+                for subkey, subvalue in value.items():
+                    flattened_dict[f"{key}_{subkey}"] = subvalue
+            else:
+                flattened_dict[key] = value
+
+        for key in flattened_dict.copy().keys():
+            if "_units" in key:
+                new_key = key.replace("_units", "/@units")
+                flattened_dict[new_key] = flattened_dict.pop(key)
+
+        for key in flattened_dict.copy().keys():
+            if key in KEYS_WITH_UNITS and f"{key}/@units" not in flattened_dict:
+                flattened_dict[f"{key}/@units"] = UNIT_MISSING[key]
```

Comment:
Reading the same dict again and again could make you algorithm slower. It is just a suggestion.
```suggestion
        def setup_unit(flattened_dict, unit_k):
            if "_units" in unit_k:
                new_key = unit_k.replace("_units", "/@units")
                flattened_dict[new_key] = flattened_dict.pop(unit_k)
            elif unit_k in KEYS_WITH_UNITS and not flattened_dict.get(f"{uni_k}/@units", None):
                flattened_dict[f"{uni_k}/@units"] = UNIT_MISSING[unit_k]

        for key, value in self.metadata.dict().items():
            if isinstance(value, dict):
                for subkey, subvalue in value.items():
                    full_sub_key = f"{key}_{subkey}"
                    flattened_dict[full_sub_key] = subvalue
                    setup_unit(flattened_dic, full_sub_key)
            else:
                flattened_dict[key] = value
                setup_unit(flattened_dic, key)
```

Reply:
used with slight modifications